core/src/main/java/com/datastax/oss/driver/internal/core/os/JnrNativeImpl.java

```diff
@@ -0,0 +1,154 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core.os;
+
+import com.datastax.oss.driver.internal.core.util.Reflection;
+import java.lang.reflect.Method;
+import java.util.Optional;
+import java.util.function.Supplier;
+import jnr.ffi.Platform;
+import jnr.posix.POSIX;
+import jnr.posix.POSIXFactory;
+import jnr.posix.Timeval;
+import jnr.posix.util.DefaultPOSIXHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JnrNativeImpl implements NativeImpl {
+
+  private static final Logger LOG = LoggerFactory.getLogger(JnrNativeImpl.class);
+
+  private final Optional<POSIX> posix;
+  private final Optional<Platform> platform;
+
+  public JnrNativeImpl() {
+
+    this.posix = loadPosix();
+    this.platform = loadPlatform();
+  }
+
+  private Optional<POSIX> loadPosix() {
+
+    try {
+      return Optional.of(POSIXFactory.getPOSIX(new DefaultPOSIXHandler(), true))
+          .flatMap(this::validatePosix);
+    } catch (Throwable t) {
+      LOG.debug("Error loading POSIX", t);
+      return Optional.empty();
+    }
+  }
+
+  private Optional<POSIX> validatePosix(POSIX posix) {
+
+    try {
+
+      posix.getpid();
+    } catch (Throwable t) {
+
+      LOG.debug("Error calling getpid()", t);
+      return Optional.empty();
+    }
+
+    try {
+
+      Timeval tv = posix.allocateTimeval();
+      int rv = posix.gettimeofday(tv);
+      if (rv != 0) {
+
+        LOG.debug("Expected getitimeofday() to return zero, observed {}", rv);
+        return Optional.empty();
+      }
+    } catch (Throwable t) {
+
+      LOG.debug("Error calling gettimeofday()", t);
+      return Optional.empty();
+    }
+
+    return Optional.of(posix);
+  }
+
+  private Optional<Platform> loadPlatform() {
+
+    try {
+
+      Class<?> platformClass = Reflection.loadClass(null, "jnr.ffi.Platform");
+      if (platformClass != null) {
+        Method getNativePlatform = platformClass.getMethod("getNativePlatform");
+        return Optional.of((jnr.ffi.Platform) getNativePlatform.invoke(null));
+      }
+      return Optional.empty();
+    } catch (Throwable t) {
+      LOG.debug("Error loading jnr.ffi.Platform class, this class will not be available.", t);
+      return Optional.empty();
+    }
+  }
+
+  @Override
+  public boolean gettimeofdayAvailable() {
+    return posix.isPresent();
+  }
+
+  @Override
+  public long gettimeofday() {
+
+    Supplier<IllegalStateException> exceptionSupplier =
+        () -> {
+          throw new IllegalStateException(
+              "Native call not available. "
+                  + "Check isCurrentTimeMicrosAvailable() before calling this method.");
+        };
+
+    Timeval tv = this.posix.map(POSIX::allocateTimeval).orElseThrow(exceptionSupplier);
+    int rv = this.posix.map(p -> p.gettimeofday(tv)).orElseThrow(exceptionSupplier);
```

Comment:
I'm being conservative here.  It's hard to imagine a case in which we'd be able to allocate a new Timeval but somehow get an exception (as opposed to a non-zero return val) when we go to make the actual syscall... but you never know.